### ISSUE TITLE
Possible to perform actions and reset sim with solute menu open

### ISSUE THREAD

**Issue by KatieWoe** [phetsims/molarity#213]
**Device**
iPad 6th generation
**OS**
iPadOS 13.3
**Browser**
Safari
**Problem Description**
For https://github.com/phetsims/QA/issues/464. Occurs when using VO on the iPad.
It is possible to use VO to leave the solute menu without closing it and perform other actions while it is open. The most noticeable it resetting the sim. This resets most things, but does not close that menu.
**Steps to Reproduce**
1. Turn on VO
2. Swipe to the solute menu and open it
3. Swipe through the options and continue swiping when you reach the end
4. When over the reset button, double click (helps to see if you've changed the state before this)

**Visuals**
![possibletoresetmenuopen](https://user-images.githubusercontent.com/41024075/71926394-73e8c200-3150-11ea-88bd-e9e09629b1b1.gif)

**Comment by KatieWoe** [comment]
Works both forwards and backwards.

**Comment by zepumph** [comment]
@jessegreenberg @terracoda, this is the first sim to use a combobox, I wonder if we ever talked about the modality of the combobox. Should we be hiding other accessible views while it is open (more like a dialog than a sim component?

**Comment by KatieWoe** [comment]
Note that I couldn't find a way to do this on computer, just iPad from what I saw.

**Comment by terracoda** [comment]
@jessegreenberg, ideally the popped-up listbox should close when it loses focus.

**Comment by terracoda** [comment]
@zepumph, good point, the combobox pattern has not been thought through yet for mobile accessibility.

**Comment by terracoda** [comment]
@zepumph, I was thinking the same thing, treat it more like a dialog, and prevent access to the rest of the sim when it is open (has focus or any of its items has focus).

**Comment by zepumph** [comment]
>@jessegreenberg, ideally the popped-up listbox should close when it loses focus.

I like this idea! It seems simple and complete.

**Comment by jessegreenberg** [comment]
>  ideally the popped-up listbox should close when it loses focus.

That sounds good to me too, Ill add that to ComboBox. This is most consistent with https://www.w3.org/TR/wai-aria-practices/examples/listbox/listbox-collapsible.html which our ComboBox is based on. For iOS this means that it will close once the user has clicked on a button or something outside of the box.

**Comment by zepumph** [comment]
I like this, and also think that this is preferable to hiding other content when the list box is open (like a dialog).

**Comment by terracoda** [comment]
There might be some details here, too: https://phetsims.github.io/binder/#sun-ComboBox

**Comment by terracoda** [comment]
Oh, I see in that doc, it says, "gesture to do" :-)

**Comment by terracoda** [comment]
@jessegreenberg, let me know if you need any help from me on this.

**Comment by jessegreenberg** [comment]
I made https://github.com/phetsims/sun/issues/557 for this since, fix proposed there.

**Comment by terracoda** [comment]
@jessegreenberg, I was playing with Molarity 
https://phet-dev.colorado.edu/html/molarity/1.6.0-dev.1/phet/molarity_en_phet.html?a11y

on my iPhone with VoiceOver on, and the combobox didn’t collapse even after a reset. 

Is 1.6.0-dev.1 the latest version?

If not, can you make a dev version with the latest changes (loss of focus closes the popped-up list of Solutes)?

And I will test again. 
Behaviour exhibited in Version 1.6.0-dev.1 is not desirable, so I just want to make sure I am am experiencing the latest and greatest version.

**Comment by jessegreenberg** [comment]
Sorry @terracoda, 1.6.0-dev.1 does not include this change. Here is a latest dev version where the combo box closes when focus leaves it:
https://phet-dev.colorado.edu/html/molarity/1.6.0-dev.2/phet/molarity_en_phet.html

**Comment by jessegreenberg** [comment]
Status update for this issue:
We can either have the behavior such that the list box traps focus/virtual cursor so that user cannot interact with other things while the list box is open, or we can allow virtual cursor to move outside of list box (but close when focus changes) and allow user interaction outside of box.

The second option is how it is currently implemented.

We found examples of both options in iOS and and ARIA examples, so we aren't sure which is best.

We ran the idea by a colleague in web a11y who thought the second option made sense because trapping virtual cursor is usually only done for dialogs. And there isn't an explicit recommendation to trap focus/virtual cursor. But he thought both choices were accessible.

@terracoda asked the W3C WAI group if they have any thoughts about this, we will also see if that provides us with any guidance.

**Comment by terracoda** [comment]
@jessegreenberg, I haven't got feedback yet. I will test the latest version today and post my thoughts.

**Comment by terracoda** [comment]
@jessegreenberg, I tested the new link and it works much better. And I also see better now the wider context of the issue. 

When the combobox is open, I have to swipe horizontally to move to a new Solute. This works great. Swiping up and down, however, moves to the next heading in the pdom moving the virtual cursor outside the open listbox and into the surrounding PDOM content. It seems like the only way to get your focus back on an item in the list is to tap on a Solute. This might be difficult and not immediately apparent to a blind user.

@jessegreenberg, is it possible to force the up/down swipes to behave the same as the horizontal swipes when the listbox is open? Is that the same as treating the listbox like a dialog?

The short of it is, I don't think we can publish as-is. I think this would be an issue for blind users.

So, I am just exploring the idea of listening for all 4 directions (vertical and horizontal swipes) to move to the next/previous solute.

**Comment by terracoda** [comment]
@jessegreenberg, if you/we can't move solute with vertical swipes, I think we will need to go to the more restrictive dialog solution.

**Comment by jessegreenberg** [comment]
>  is it possible to force the up/down swipes to behave the same as the horizontal swipes when the listbox is open?

No, this would not be possible but this would also not be better for the user. The behavior you described is just how VoiceOver works.  Left/Right swipes move to previous/next items. Up/down swipes move to previous/next items *as set by your rotor*. So if set to "headings" it will move between them, and that is the expected behavior.

Do you still think we should go with a more restrictive solution?

**Comment by terracoda** [comment]
@jessegreenberg, I think so. I think it could be very confusing to have to touch an item in the popped up listbox to get focus back in there if focus is moved outside the box with an up/down gesture.

Unless, I am missing something? I don't know how to get focus back on the popped up list without touching it directly.

Do you have any thoughts?

**Comment by terracoda** [comment]
And I was lookig around for VO gesture advice for listboxes and comboboxes for a bit, and can't seem to find anything specific. I'll reach out to the list one more time with simpler question.

Advice or expected behaviour for VO gesture support for a popped up list box or combobox.

**Comment by jessegreenberg** [comment]
> I think it could be very confusing to have to touch an item in the popped

That is not the only way, you can also use normal left/right gestures to move virtual cursor back into the combo box if cursor leaves it.

**Comment by terracoda** [comment]
Just an FYI to myself... I noticed and posted on slack:
> Oh, I see, yes I can either flick left/right or up/down to go outside the box. At the top or bottom of the popped-up list, neither keyboard or flicks cycle through the items. In the PhET menu, for keyboard, the arrow keys cycle through, but for gesture, it doesn't. We have the same issue there, the cursor can move outside the popped up menu.

**Comment by terracoda** [comment]
Over the past couple of days, I got a several responses from people on the public-ARIA list and the W3C-WAI-IG list. @SinaBahram, Matt King, @marc_haunschild on Molarity's combobox interaction (a.k.a button that has popped-up listbox).

An important, but unrelated comment from @SinaBahram was that "the Unicode characters everywhere which make the speech very frustrating to listen to". I'll post a separate issue about that as I think it is already being addressed.

Two awesome articles were shared:
- https://www.24a11y.com/2019/select-your-poison/
- https://www.24a11y.com/2019/select-your-poison-part-2/

And Matt mentioned changes coming to `combobox` role. The new design pattern is covered in the articles above. Our current design pattern, I think is based on ARIA 1.1:
- https://www.w3.org/TR/wai-aria-practices-1.2/#combobox

I am wondering if there is anyway we can make the solute scrolling flicks stay in the box and cycle through the list the way it works with the keyboard?

**Comment by terracoda** [comment]
@jessegreenberg and @terracoda discussed the resources, ideas, and possibilities that were brought up in the previous comment.

We feel that while our current implementation is accessible on iOS with VO enabled, it is not optimal as it behaves differently from the native `select` would, specifically, a user can easily flick in and out of the popped up list of options without the list of options closing.

We decided to explore the possibility of using a dialog-like experience for the popped-up list of options. The intention is to keep the user in the box until the make a selection. Specifically this means:

- a double tap on the focused button pops-up the list of solutes, and the rest of the pdom become hidden thus making it inaccessible with vertical flicks.
- horizontal flicks move through the popped-up list of solutes - right flicks move down the list of solutes, and left flicks move up the list of solutes.
- double tap selects the solute that has focus (i.e., the solute the user has moved to) and closes the popped-up list. And an object response fires announcing the the changed solute (Nickel II Chloride Solute, pop-up button), followed immediately by a context response describing the state of the new solute. For example: "Green solution, low concentration." 
- vertical flicks do nothing as the pdom is hidden. The user must make a selection to close the popped-up list of options.

**Note** there might be a native gesture that dismisses the popped-up list without making a selection, but I haven't found one yet.
- https://www.applevis.com/guides/complete-list-ios-and-ipados-gestures-available-voiceover-users
- https://www.freedomscientific.com/Content/Documents/Other/JAWSTouchSupportOverview.htm

**Note** that I have noticed other native reading gestures work nicely with our implementation like the gesture to read from the beginning or to read from where you currently are.

**Comment by terracoda** [comment]
@jessegreenberg and I briefly met and our dialog-ides for the listbox won't work because of the way Accessible Order currently works in Scenery. 

We can hide the rest of the PDOM when a dialog is popped up because the dialogs are on top of everything else. However, with the combobox in Molarity, the popped-up list has content layers above and below it in the Accessible Order. It is not trivial (not possible at this time) to trap the virtual cursor in the popped-up list like we do for dialogs. Non-trivial changes to Scenery would have to occur for this to be possible. We think this might be useful but not necessary to hold up the publication of Molarity.

The Molarity sim and its Solute combobox interaction are accessible and usable with VO gestures on iOS. The listbox can get left open if a selection is not made when using iOS on VO, and it can also easily be re-entered using the virtual cursor if the box gets left open. It doesn't seem to be a barrier, though perhaps visually not like the way the other platforms behave.

Marking as won't fix for now, and closing.

**Comment by jessegreenberg** [comment]
Thanks for summarizing @terracoda. Im going to keep this open as a reminder because changes from https://github.com/phetsims/sun/issues/557 still need to go into the release branch.

**Comment by terracoda** [comment]
Yes, the original issues are fixed.

**Comment by KatieWoe** [comment]
This sort of thing also happens with the PhET menu and was seen in https://github.com/phetsims/QA/issues/471

**Comment by jessegreenberg** [comment]
Thanks @KatieWoe that makes a lot of sense. I verified that the PhET Menu closed when activating a button outside of it with iOS VoiceOver, which is correct behavior. This is acceptable behavior for now. If we address https://github.com/phetsims/scenery/issues/1033 we will be better equipped to trap the virtual cursor inside menus.

**Comment by terracoda** [comment]
@KatieWoe, thanks for adding the co

**Comment by zepumph** [comment]
I am going to proceed with https://github.com/phetsims/QA/issues/471 without taking any commits from this issue.

**Comment by zepumph** [comment]
@jessegreenberg in lieu of doing https://github.com/phetsims/scenery/issues/1033, do you think that we could make the listBoxParent `sim.topLayer` or something? That is a pretty fragile choice, but perhaps there is something that we could do to put place this outside of the screen hierarchy. Likely we could calculate its positioning and wouldn't need to disrupt much. Here is some sketchy code that doesn't totally work:

```js

Index: js/ComboBox.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/ComboBox.js	(revision 1445d512af968a175a4164d4175443502d0cb323)
+++ js/ComboBox.js	(date 1580881158842)
@@ -317,6 +317,9 @@
         this.scaleListBox();
         this.moveListBox();
         this.listBox.moveToFront();
+        this.listBox.detach();
+        phet.joist.sim.topLayer.addChild( this.listBox );
+        this.listBox.center = this.listBox.localToGlobalPoint( this.listBox.center );
         this.listBox.visible = true;
 
         // manage clickToDismissListener
@@ -342,6 +345,11 @@
           this.display = null;
         }
 
+        this.listBox.detach();
+        this.listParent.addChild( this.listBox );
+        this.listBox.center = this.listParent.globalToLocalPoint( this.listBox.localToGlobalPoint( this.listBox.center ) );
+
+
         // hide the list box
         this.listBox.visible = false;
 
```

If I could get the transform to work, then you should be able to set the accessible views visible false
on the screen since the listBox parent isn't in the screen anymore. I have a bad feeling about this, but thought I should put it out there anyways.

**Comment by jessegreenberg** [comment]
That seems reasonable. I would like to ask the author of ComboBox why listParent is required instead of putting on top layer. Maybe layering it under something was important for a sim.

**Comment by jessegreenberg** [comment]
Instead of adding directly to topLayer, I think the next step is to use sim.showPopup. Unsure if that will work at this time.  See above issue.

I am moving this to https://github.com/phetsims/sun/issues/565 and closing the molarity report.

**Comment by KatieWoe** [comment]
rc.3 seems to be behaving as expected
